primitives/state-machine/src/trie_backend_essence.rs

```diff
@@ -80,9 +103,20 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
 
 	/// Set trie root. This is useful for testing.
 	pub fn set_root(&mut self, root: H::Out) {
+		// if root did change so can have child roots.
+		self.reset_child_roots_cache();
 		self.root = root;
 	}
 
+	#[cfg(feature = "std")]
+	fn reset_child_roots_cache(&mut self) {
+		self.cache = Arc::new(RwLock::new(Cache::new()));
```

Comment:
This also doesn't reflect the name of the function. If this should only reset the child roots, you should do it. Currently you clear the entire cache.